README.md
weather.wundergroundpws

A Kodi weather addon for registered and active Weather Underground personal weather station users.

:+1: If you find this product useful, feel free to buy me a beer: https://paypal.me/cytecheng

To use this addon, you need to enter your personal weather station API key in the addon settings.

To get a free API key:
1) You must have a personal weather station registered and uploading data to Weather Underground
    
    a) Join weather Underground
    
    b) Sign In
    
    c) My Profile -> My Weather Stations
    
    d) Add a New PWS
2) get API key at  https://www.wunderground.com/member/api-keys
3) In Kodi - System - Addons - Install from Zip File (weather.wundergroundpws.zip)
4) In Kodi - System - Services - Weather select Weather UndergroundPWS as Service for weather information
5) -Settings add your API Key
6) -Settings

    Location1 Display Name - ( This can be anything you want)
    
    Location1 StationID - (your station ID  i.e  KCAFOOBARFXX)
    
    Location1 Forecast Latitude - ( 35.317)
    
    Location1 Forecast Longitude - ( -119.125)
    
The StationID collects the data from the PWS and displays in the current observation (top area) in Kodi Weather.

The latitude/longitude defines where the forecast data (Kodi 5 day forecast) is collected

Only English and English(US) languages are defined. All others will fail.

This addon no longer allows for geoip lookup.
You must know the stationid and forecast latitude/longitude beforehand to configure.



changelog.txt
v1.0.4
-fix error with winddir and humidity when wunderground returns null
v1.0.3
-fix error with winddir and humidity when wunderground returns null
v1.0.2
-corrected pressure, precipRate, heatIndex and windChill conversions (skin dependent properties)
v1.0.1
-corrected null values returned from WU api
from weather company API doc:
    PLEASE NOTE: The daypart object (all keys[0]) as well as the temperatureMax field OUTSIDE of the daypart object
    will appear as null in the API after 3:00pm Local Apparent Time.
-set iconCode on key[0] to nightime code and
-temperature of null to 'na'

v1.0.0
Rework old Wunderground addon by Ronie v6.0.1
- Weather Underground personal weather stations only via
  free Weather Company API for PWS uploaders


resources/lib/utilities.py
# -*- coding: utf-8 -*-

from __future__ import division
from builtins import str
from past.utils import old_div
import sys, math
import xbmc

LANGUAGE = sys.modules[ "__main__" ].LANGUAGE

#http://www.wunderground.com/weather/api/d/docs?d=language-support
        # kodi lang name         # wu code
LANG = { 'afrikaans'             : 'AF',
         'albanian'              : 'AL',
         'amharic'               : 'EN', # AM is n/a, use AR or EN?
         'arabic'                : 'AR',
         'armenian'              : 'HY',
         'azerbaijani'           : 'AZ',
         'basque'                : 'EU',
         'belarusian'            : 'BY',
         'bosnian'               : 'CR', # BS is n/a, use CR or SR?
         'bulgarian'             : 'BU',
         'burmese'               : 'MY',
         'catalan'               : 'CA',
         'chinese (simple)'      : 'CN',
         'chinese (traditional)' : 'TW',
         'croatian'              : 'CR',
         'czech'                 : 'CZ',
         'danish'                : 'DK',
         'dutch'                 : 'NL',
         'english'               : 'LI',
         'english (us)'          : 'EN',
         'english (australia)'   : 'LI',
         'english (new zealand)' : 'LI',
         'esperanto'             : 'EO',
         'estonian'              : 'ET',
         'faroese'               : 'DK', # FO is n/a, use DK
         'finnish'               : 'FI',
         'french'                : 'FR',
         'galician'              : 'GZ',
         'german'                : 'DL',
         'greek'                 : 'GR',
         'georgian'              : 'KA',
         'hebrew'                : 'IL',
         'hindi (devanagiri)'    : 'HI',
         'hungarian'             : 'HU',
         'icelandic'             : 'IS',
         'indonesian'            : 'ID',
         'italian'               : 'IT',
         'japanese'              : 'JP',
         'korean'                : 'KR',
         'latvian'               : 'LV',
         'lithuanian'            : 'LT',
         'macedonian'            : 'MK',
         'malay'                 : 'EN', # MS is n/a, use EN
         'malayalam'             : 'EN', # ML is n/a, use EN
         'maltese'               : 'MT',
         'maori'                 : 'MI',
         'mongolian (mongolia)'  : 'MN',
         'norwegian'             : 'NO',
         'ossetic'               : 'EN', # OS is n/a, use EN
         'persian'               : 'FA',
         'persian (iran)'        : 'FA',
         'polish'                : 'PL',
         'portuguese'            : 'BR',
         'portuguese (brazil)'   : 'BR',
         'romanian'              : 'RO',
         'russian'               : 'RU',
         'serbian'               : 'SR',
         'serbian (cyrillic)'    : 'SR',
         'silesian'              : 'PL', # SZL is n/a, use PL or EN?
         'sinhala'               : 'EN', # SI is n/a, use EN?
         'slovak'                : 'SK',
         'slovenian'             : 'SL',
         'spanish'               : 'SP',
         'spanish (argentina)'   : 'SP',
         'spanish (mexico)'      : 'SP',
         'swedish'               : 'SW',
         'tajik'                 : 'FA', # TG is n/a, use FA or EN?
         'tamil (india)'         : 'EN', # TA is n/a, use EN
         'telugu'                : 'EN', # TE is n/a, use EN
         'thai'                  : 'TH',
         'turkish'               : 'TU',
         'ukrainian'             : 'UA',
         'uzbek'                 : 'UZ',
         'vietnamese'            : 'VU',
         'welsh'                 : 'CY'}

                # WU    KODI    description
WEATHER_CODES = {0: '0',  # Tornado	Forecast	Night + Day
                 1: '1',  # Tropical Storm	Forecast + Observations	Night + Day
                 2: '2',  # Hurricane	Forecast	Night + Day
                 3: '3',  # Strong Storms	Forecast	Night + Day
                 4: '4',  # Thunder and Hail	Forecast + Observations	Night + Day
                 5: '5',  # Rain to Snow Showers	Forecast + Observations	Night + Day
                 6: '6',  # Rain / Sleet	Forecast + Observations	Night + Day
                 7: '7',  # Wintry Mix Snow / Sleet	Forecast + Observations	Night + Day
                 8: '8',  # Freezing Drizzle	Forecast + Observations	Night + Day
                 9: '9',  # Drizzle	Forecast + Observations	Night + Day
                 10: '10',  # Freezing Rain	Forecast + Observations	Night + Day
                 11: '11',  # Light Rain	Forecast + Observations	Night + Day
                 12: '12',  # Rain	Forecast + Observations	Night + Day
                 13: '13',  # Scattered Flurries	Forecast + Observations	Night + Day
                 14: '14',  # Light Snow	Forecast + Observations	Night + Day
                 15: '15',  # Blowing / Drifting Snow	Forecast + Observations	Night + Day
                 16: '16',  # Snow	Forecast + Observations	Night + Day
                 17: '17',  # Hail	Forecast + Observations	Night + Day
                 18: '18',  # Sleet	Forecast + Observations	Night + Day
                 19: '19',  # Blowing Dust / Sandstorm	Forecast + Observations	Night + Day
                 20: '20',  # Foggy	Forecast + Observations	Night + Day
                 21: '21',  # Haze / Windy	Forecast + Observations	Night + Day
                 22: '22',  # Smoke / Windy	Forecast + Observations	Night + Day
                 23: '23',  # Breezy	Forecast	Night + Day
                 24: '24',  # Blowing Spray / Windy	Forecast + Observations	Night + Day
                 25: '25',  # Frigid / Ice Crystals	Forecast + Observations	Night + Day
                 26: '26',  # Cloudy	Forecast + Observations	Night + Day
                 27: '27',  # Mostly Cloudy	Forecast + Observations	Night + Day
                 28: '28',  # Mostly Cloudy	Forecast + Observations	Day
                 29: '29',  # Partly Cloudy	Forecast + Observations	Night
                 30: '30',  # Partly Cloudy	Forecast + Observations	Day
                 31: '31',  # Clear	Forecast + Observations	Night
                 32: '32',  # Sunny	Forecast + Observations	Day
                 33: '33',  # Fair / Mostly Clear	Forecast + Observations	Night
                 34: '34',  # Fair / Mostly Sunny	Forecast + Observations	Day
                 35: '35',  # Mixed Rain & Hail	Forecast	Day
                 36: '36',  # Hot	Forecast	Day
                 37: '37',  # Isolated Thunderstorms	Forecast	Day
                 38: '38',  # Thunderstorms	Forecast + Observations	Night + Day
                 39: '39',  # Scattered Showers	Forecast	Day
                 40: '40',  # Heavy Rain	Forecast + Observations	Night + Day
                 41: '41',  # Scattered Snow Showers	Forecast	Day
                 42: '42',  # Heavy Snow	Forecast + Observations	Night + Day
                 43: '43',  # Blizzard	Forecast	Night + Day
                 44: 'na',  # Not Available (N/A)	Forecast	Night + Day
                 45: '45',  # Scattered Showers	Forecast	Night
                 46: '46',  # Scattered Snow Showers	Forecast	Night
                 47: '47',  # Scattered Thunderstorms	Forecast + Observations	Night + Day
                 }

MONTH = { 1  : xbmc.getLocalizedString(51),
          2  : xbmc.getLocalizedString(52),
          3  : xbmc.getLocalizedString(53),
          4  : xbmc.getLocalizedString(54),
          5  : xbmc.getLocalizedString(55),
          6  : xbmc.getLocalizedString(56),
          7  : xbmc.getLocalizedString(57),
          8  : xbmc.getLocalizedString(58),
          9  : xbmc.getLocalizedString(59),
          10 : xbmc.getLocalizedString(60),
          11 : xbmc.getLocalizedString(61),
          12 : xbmc.getLocalizedString(62)}

WEEKDAY = { 0  : xbmc.getLocalizedString(41),
            1  : xbmc.getLocalizedString(42),
            2  : xbmc.getLocalizedString(43),
            3  : xbmc.getLocalizedString(44),
            4  : xbmc.getLocalizedString(45),
            5  : xbmc.getLocalizedString(46),
            6  : xbmc.getLocalizedString(47)}

SEVERITY = { 'W' : LANGUAGE(32510),
             'A' : LANGUAGE(32511),
             'Y' : LANGUAGE(32512),
             'S' : LANGUAGE(32513),
             'O' : LANGUAGE(32514),
             'F' : LANGUAGE(32515),
             'N' : LANGUAGE(32516),
             'L' :'', # no idea
             ''  : ''}

def MOONPHASE(age, percent):
    if (percent == 0) and (age == 0):
        phase = LANGUAGE(32501)
    elif (age < 17) and (age > 0) and (percent > 0) and (percent < 50):
        phase = LANGUAGE(32502)
    elif (age < 17) and (age > 0) and (percent == 50):
        phase = LANGUAGE(32503)
    elif (age < 17) and (age > 0) and (percent > 50) and (percent < 100):
        phase = LANGUAGE(32504)
    elif (age > 0) and (percent == 100):
        phase = LANGUAGE(32505)
    elif (age > 15) and (percent < 100) and (percent > 50):
        phase = LANGUAGE(32506)
    elif (age > 15) and (percent == 50):
        phase = LANGUAGE(32507)
    elif (age > 15) and (percent < 50) and (percent > 0):
        phase = LANGUAGE(32508)
    else:
        phase = ''
    return phase

def KPHTOBFT(spd):
    if (spd < 1.0):
        bft = '0'
    elif (spd >= 1.0) and (spd < 5.6):
        bft = '1'
    elif (spd >= 5.6) and (spd < 12.0):
        bft = '2'
    elif (spd >= 12.0) and (spd < 20.0):
        bft = '3'
    elif (spd >= 20.0) and (spd < 29.0):
        bft = '4'
    elif (spd >= 29.0) and (spd < 39.0):
        bft = '5'
    elif (spd >= 39.0) and (spd < 50.0):
        bft = '6'
    elif (spd >= 50.0) and (spd < 62.0):
        bft = '7'
    elif (spd >= 62.0) and (spd < 75.0):
        bft = '8'
    elif (spd >= 75.0) and (spd < 89.0):
        bft = '9'
    elif (spd >= 89.0) and (spd < 103.0):
        bft = '10'
    elif (spd >= 103.0) and (spd < 118.0):
        bft = '11'
    elif (spd >= 118.0):
        bft = '12'
    else:
        bft = ''
    return bft

def CELSIUStoFAHR(temp):
    if temp is None:
        return 'na'
    return 9/5 * temp + 32

def wind_deg_to_card(deg):
    if deg == "na":
        return 'na'
    else:
        arr = ['NNE', 'NE', 'ENE', 'E', 'ESE', 'SE', 'SSE', 'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW', 'N']
        return arr[int(abs((deg - 11.25) % 360) / 22.5)]

def FEELS_LIKE(T, V=0, R=0, ext=''):
    if T <= 10 and V >= 8:
        FeelsLike = WIND_CHILL(T, V)
    elif T >= 26:
        FeelsLike = HEAT_INDEX(T, R)
    else:
        FeelsLike = T
    if ext == 'F':
        FeelsLike = FeelsLike * 1.8 + 32
    return str(int(round(FeelsLike)))

#### thanks to FrostBox @ http://forum.kodi.tv/showthread.php?tid=114637&pid=937168#pid937168
def WIND_CHILL(T, V):
    FeelsLike = ( 13.12 + ( 0.6215 * T ) - ( 11.37 * V**0.16 ) + ( 0.3965 * T * V**0.16 ) )
    return FeelsLike

### https://en.wikipedia.org/wiki/Heat_index
def HEAT_INDEX(T, R):
    T = T * 1.8 + 32 # calaculation is done in F
    FeelsLike = -42.379 + (2.04901523 * T) + (10.14333127 * R) + (-0.22475541 * T * R) + (-0.00683783 * T**2) + (-0.05481717 * R**2) + (0.00122874 * T**2 * R) + (0.00085282 * T * R**2) + (-0.00000199 * T**2 * R**2)
    FeelsLike = (FeelsLike - 32) / 1.8 # convert to C
    return FeelsLike

#### thanks to FrostBox @ http://forum.kodi.tv/showthread.php?tid=114637&pid=937168#pid937168
def DEW_POINT(Tc=0, RH=93, ext='', minRH=( 0, 0.075 )[ 0 ]):
    Es = 6.11 * 10.0**( 7.5 * Tc / ( 237.7 + Tc ) )
    RH = RH or minRH
    E = old_div(( RH * Es ), 100)
    try:
        DewPoint = old_div(( -430.22 + 237.7 * math.log( E ) ), ( -math.log( E ) + 19.08 ))
    except ValueError:
        DewPoint = 0
    if ext == 'F':
        DewPoint = DewPoint * 1.8 + 32
    return str(int(round(DewPoint)))


def MBtoINHG(pressure):
    if pressure is None:
        return 'na'
    pressure = pressure * 0.02953
    return str(round(pressure, 2))


def MMtoIN(precip):
    if precip is None:
        return 'na'
    precip = precip * 0.0393701
    return str(round(precip, 2))


resources/lib/wunderground.py
# -*- coding: utf-8 -*-

from future import standard_library
standard_library.install_aliases()
import sys, urllib.request, urllib.error, urllib.parse, gzip, base64, traceback
import xbmc, xbmcgui, xbmcaddon, xbmcvfs
from io import StringIO
from io import BytesIO
import json

API = sys.modules[ "__main__" ].API
FORECAST5_URL = 'https://api.weather.com/v3/wx/forecast/daily/5day?geocode={},{}&units=m&language=en-US&format=json&apiKey={}'
CURRENT_PWS_URL = 'https://api.weather.com/v2/pws/observations/current?stationId={}&format=json&units=m&apiKey={}'

def wundergroundapi(locid, loclat, loclon):
    try:
        forecast5_url = FORECAST5_URL.format(loclat,loclon,API)
        forecast5_req = urllib.request.Request(forecast5_url)
        forecast5_req.add_header('Accept-encoding', 'gzip')
        response = urllib.request.urlopen(forecast5_req)
        if response.info().get('Content-Encoding') == 'gzip':
            buf = BytesIO(response.read())
            compr = gzip.GzipFile(fileobj=buf)
            forecast5_data = compr.read()
        else:
            forecast5_data = response.read()
        response.close()
    except:
        forecast5_data = ''

    try:
        currentpws_url = CURRENT_PWS_URL.format(locid, API)
        currentpws_req = urllib.request.Request(currentpws_url)
        currentpws_req.add_header('Accept-encoding', 'gzip')
        response = urllib.request.urlopen(currentpws_req)
        if response.info().get('Content-Encoding') == 'gzip':
            buf = BytesIO(response.read())
            compr = gzip.GzipFile(fileobj=buf)
            dcurrentpws_ata = compr.read()
        else:
            currentpws_data = response.read()
        response.close()
    except:
        currentpws_data = ''

    #this is python 3.5
    #jsonMerged = {**json.loads(currentpws_data), **json.loads(forecast5_data)}

    a = json.loads(currentpws_data)
    b = json.loads(forecast5_data)
    jsonMerged = merge_two_dicts(a, b)

    data = json.dumps(jsonMerged)

    return data

def merge_two_dicts(x, y):
    """Given two dicts, merge them into a new dict as a shallow copy."""
    z = x.copy()
    z.update(y)
    return z


default.py
# -*- coding: utf-8 -*-

# *  This Program is free software; you can redistribute it and/or modify
# *  it under the terms of the GNU General Public License as published by
# *  the Free Software Foundation; either version 2, or (at your option)
# *  any later version.
# *
# *  This Program is distributed in the hope that it will be useful,
# *  but WITHOUT ANY WARRANTY; without even the implied warranty of
# *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# *  GNU General Public License for more details.
# *
# *  You should have received a copy of the GNU General Public License
# *  along with Kodi; see the file COPYING. If not, write to
# *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
# *  http://www.gnu.org/copyleft/gpl.html


# Kodi expects a weather addon to return these properties:
#
# * Current.Condition (the weather forecast)
# * Current.Temperature (temperature in degrees celcius)
# * Current.Wind (windspeed in km/h)
# * Current.WindDirection (wind direction, NNE)
# * Current.Humidity (humidity in %)
# * Current.FeelsLike (feels-like temperature in degrees celcius)
# * Current.UVIndex (the uv-index)
# * Current.DewPoint (dewpoint in degrees celcius)
# * Current.OutlookIcon (weather icon, 34.png)
# * Current.FanartCode (condition code, 34)
# * Day%i.Title (name of the day)
# * Day%i.HighTemp (high temp of the day in degrees celcius)
# * Day%i.LowTemp (low temp of the day in degrees celcius)
# * Day%i.Outlook (the weather forecast)
# * Day%i.OutlookIcon (weather icon, 18.png)
# * Day%i.FanartCode (condition code, 18)

from __future__ import division
from future import standard_library
standard_library.install_aliases()
from builtins import str
from builtins import range
from past.utils import old_div
import os, sys, socket, unicodedata, urllib.request, urllib.error, urllib.parse, time, gzip
import datetime
from io import StringIO
import xbmc, xbmcgui, xbmcaddon, xbmcvfs
import json

ADDON      = xbmcaddon.Addon()
ADDONNAME  = ADDON.getAddonInfo('name')
ADDONID    = ADDON.getAddonInfo('id')
CWD        = ADDON.getAddonInfo('path').decode("utf-8")
VERSION    = ADDON.getAddonInfo('version')
LANGUAGE   = ADDON.getLocalizedString
RESOURCE   = xbmc.translatePath( os.path.join( CWD, 'resources', 'lib' ).encode("utf-8") ).decode("utf-8")
PROFILE    = xbmc.translatePath(ADDON.getAddonInfo('profile')).decode('utf-8')
API        = ADDON.getSetting('API')
STATIONNAME= ADDON.getSetting('Location1')
STATIONID  = ADDON.getSetting('Location1id')
FORECASTLAT= ADDON.getSetting('Location1lat')
FORECASTLON= ADDON.getSetting('Location1lon')

sys.path.append(RESOURCE)

from utilities import *
from wunderground import wundergroundapi

FORMAT           = 'json'
DEBUG            = ADDON.getSetting('Debug')
XBMC_PYTHON      = xbmcaddon.Addon(id='xbmc.python').getAddonInfo('version')
WEATHER_ICON     = '%s.png'
WEATHER_WINDOW   = xbmcgui.Window(12600)
LOCALIZE         = xbmc.getLanguage().lower()
SPEEDUNIT        = xbmc.getRegion('speedunit')
TEMPUNIT         = str(xbmc.getRegion('tempunit'),encoding='utf-8')
TIMEFORMAT       = xbmc.getRegion('meridiem')
DATEFORMAT       = xbmc.getRegion('dateshort')
MAXDAYS          = 5

socket.setdefaulttimeout(10)

def log(txt):
    if DEBUG == 'true':
        if isinstance (txt,str):
            txt = txt.decode("utf-8")
        message = u'%s: %s' % (ADDONID, txt)
        xbmc.log(msg=message.encode("utf-8"), level=xbmc.LOGDEBUG)

def set_property(name, value):
    WEATHER_WINDOW.setProperty(name, value)

def refresh_locations():
    locations = 0
    for count in range(1, 6):
        loc_name = ADDON.getSetting('Location%s' % count)
        if loc_name != '':
            locations += 1
        else:
            ADDON.setSetting('Location%sid' % count, '')
        set_property('Location%s' % count, loc_name)
    set_property('Locations', str(locations))
    log('available locations: %s' % str(locations))

def forecast(loc,locid, loclat, loclon):
    try:
        lang = LANG[LOCALIZE]
    except:
        lang = 'EN'
    opt = 'lang:' + lang
    log('weather location: %s' % locid)
    retry = 0
    while (retry < 6) and (not xbmc.abortRequested):
        query = wundergroundapi(locid, loclat, loclon)

        if query != '':
            retry = 6
        else:
            retry += 1
            xbmc.sleep(10000)
            log('weather download failed')
    log('forecast data: %s' % query)
    data = parse_data(query)
    if data != '' and 'observations' in data and 'error' not in data['observations']:
        properties(data,loc,locid)
    else:
        clear()

def clear():
    set_property('Current.Condition'     , 'N/A')
    set_property('Current.Temperature'   , '0')
    set_property('Current.Wind'          , '0')
    set_property('Current.WindDirection' , 'N/A')
    set_property('Current.Humidity'      , '0')
    set_property('Current.FeelsLike'     , '0')
    set_property('Current.UVIndex'       , '0')
    set_property('Current.DewPoint'      , '0')
    set_property('Current.OutlookIcon'   , 'na.png')
    set_property('Current.FanartCode'    , 'na')
    for count in range (0, MAXDAYS+1):
        set_property('Day%i.Title'       % count, 'N/A')
        set_property('Day%i.HighTemp'    % count, '0')
        set_property('Day%i.LowTemp'     % count, '0')
        set_property('Day%i.Outlook'     % count, 'N/A')
        set_property('Day%i.OutlookIcon' % count, 'na.png')
        set_property('Day%i.FanartCode'  % count, 'na')

def parse_data(response):
    try:
        raw = response.replace('<br>',' ').replace('&auml;','ä') # wu api bugs
        #reply = raw.replace('"-999%"','""').replace('"-9999.00"','""').replace('"-9998"','""').replace('"NA"','""').replace('"--"','""') # wu will change these to null responses in the future
        reply = raw.replace('null', '"na"')
        data = json.loads(reply)
    except:
        log('failed to parse weather data')
        data = ''
    return data

def properties(data,loc,locid):
# standard properties
    set_property('Current.Location'      , str(data['observations'][0]['stationID']))
    set_property('Current.Temperature'   , str(data['observations'][0]['metric']['temp']))
    set_property('Current.Wind'          , str(data['observations'][0]['metric']['windSpeed']))
    set_property('Current.WindDirection' , str(wind_deg_to_card(data['observations'][0]['winddir'])))
    set_property('Current.Humidity'      , str(data['observations'][0]['humidity']))
    set_property('Current.FeelsLike' , FEELS_LIKE(int(data['observations'][0]['metric']['temp']), data['observations'][0]['metric']['windSpeed'], data['observations'][0]['humidity'], 'c') + TEMPUNIT)
    set_property('Current.UVIndex'       , str(data['observations'][0]['uv']))
    set_property('Current.DewPoint'      , str(data['observations'][0]['metric']['dewpt']))

# forecast properties
    set_property('Forecast.IsFetched'        , 'true')
    set_property('Forecast.City'             , "")
    set_property('Forecast.State'            , "")
    set_property('Forecast.Country'          , "")
    if data['observations'][0]['epoch']:
        update = time.localtime(float(data['observations'][0]['epoch']))
        local = time.localtime(float(data['observations'][0]['epoch']))
        if DATEFORMAT[1] == 'd':
            updatedate = WEEKDAY[update[6]] + ' ' + str(update[2]) + ' ' + MONTH[update[1]] + ' ' + str(update[0])
            localdate = WEEKDAY[local[6]] + ' ' + str(local[2]) + ' ' + MONTH[local[1]] + ' ' + str(local[0])
        elif DATEFORMAT[1] == 'm':
            updatedate = WEEKDAY[update[6]] + ' ' + MONTH[update[1]] + ' ' + str(update[2]) + ', ' + str(update[0])
            localdate = WEEKDAY[local[6]] + ' ' + str(local[2]) + ' ' + MONTH[local[1]] + ' ' + str(local[0])
        else:
            updatedate = WEEKDAY[update[6]] + ' ' + str(update[0]) + ' ' + MONTH[update[1]] + ' ' + str(update[2])
            localdate = WEEKDAY[local[6]] + ' ' + str(local[0]) + ' ' + MONTH[local[1]] + ' ' + str(local[2])
        if TIMEFORMAT != '/':
            updatetime = time.strftime('%I:%M%p', update)
            localtime = time.strftime('%I:%M%p', local)
        else:
            updatetime = time.strftime('%H:%M', update)
            localtime = time.strftime('%H:%M', local)
        set_property('Forecast.Updated'          , updatedate + ' - ' + updatetime)
# current properties
    set_property('Current.LocalTime'         , localtime)
    set_property('Current.LocalDate'         , localdate)
    set_property('Current.IsFetched'         , 'true')
    set_property('Current.WindDegree'        , str(data['observations'][0]['winddir']) + u'°')
    set_property('Current.SolarRadiation'    , str(data['observations'][0]['uv']))
    if 'F' in TEMPUNIT:
        set_property('Current.Pressure'      , str(MBtoINHG(data['observations'][0]['metric']['pressure'])) + ' inHg')
        set_property('Current.Precipitation' , str(MMtoIN(data['observations'][0]['metric']['precipRate'])) + ' in/hr')
        set_property('Current.HeatIndex'     , str(CELSIUStoFAHR(data['observations'][0]['metric']['heatIndex'])) + TEMPUNIT)
        set_property('Current.WindChill'     , str(CELSIUStoFAHR(data['observations'][0]['metric']['windChill'])) + TEMPUNIT)
    else:
        set_property('Current.Pressure'      , str(data['observations'][0]['metric']['pressure']) + ' mb')
        set_property('Current.Precipitation' , str(data['observations'][0]['metric']['precipRate']) + ' mm/hr')
        set_property('Current.HeatIndex'     , str(data['observations'][0]['metric']['heatIndex']) + TEMPUNIT)
        set_property('Current.WindChill'     , str(data['observations'][0]['metric']['windChill']) + TEMPUNIT)
    if SPEEDUNIT == 'mph':
        set_property('Current.WindGust'      , str(data['observations'][0]['metric']['windGust']) + ' ' + SPEEDUNIT)
    else:
        set_property('Current.WindGust'      , str(data['observations'][0]['metric']['windGust']) + ' ' + SPEEDUNIT)
# today properties
    set_property('Today.IsFetched'                     , 'true')
    if TIMEFORMAT != '/':
        AM = str(TIMEFORMAT.split('/')[0],encoding='utf-8')
        PM = str(TIMEFORMAT.split('/')[1],encoding='utf-8')
        hour = int(datetime.datetime.fromtimestamp(int(data['sunriseTimeUtc'][0])).strftime('%H')) % 24
        isam = (hour >= 0) and (hour < 12)
        if isam:
            hour = ('12' if (hour == 0) else '%02d' % (hour))
            set_property('Today.Sunrise'               , hour.lstrip('0') + ':' + datetime.datetime.fromtimestamp(int(data['sunriseTimeUtc'][0])).strftime('%M') + ' ' + AM)
        else:
            hour = ('12' if (hour == 12) else '%02d' % (hour-12))
            set_property('Today.Sunrise'               , hour.lstrip('0') + ':' + datetime.datetime.fromtimestamp(int(data['sunriseTimeUtc'][0])).strftime('%M') + ' ' + PM)
        hour = int(datetime.datetime.fromtimestamp(int(data['sunsetTimeUtc'][0])).strftime('%H')) % 24
        isam = (hour >= 0) and (hour < 12)
        if isam:
            hour = ('12' if (hour == 0) else '%02d' % (hour))
            set_property('Today.Sunset'                , hour.lstrip('0') + ':' + datetime.datetime.fromtimestamp(int(data['sunsetTimeUtc'][0])).strftime('%M') + ' ' + AM)
        else:
            hour = ('12' if (hour == 12) else '%02d' % (hour-12))
            set_property('Today.Sunset'                , hour.lstrip('0') + ':' + datetime.datetime.fromtimestamp(int(data['sunsetTimeUtc'][0])).strftime('%M') + ' ' + PM)
    else:
        set_property('Today.Sunrise'                   , datetime.datetime.fromtimestamp(int(data['sunriseTimeUtc'][0])).strftime('%H') + ':' + datetime.datetime.fromtimestamp(int(data['sunriseTimeUtc'][0])).strftime('%M'))
        set_property('Today.Sunset'                    , datetime.datetime.fromtimestamp(int(data['sunsetTimeUtc'][0])).strftime('%H') + ':' + datetime.datetime.fromtimestamp(int(data['sunriseTimeUtc'][0])).strftime('%M'))
# daily properties
    set_property('Daily.IsFetched', 'true')
    for count, item in enumerate(data['dayOfWeek']):
        # PLEASE NOTE: The daypart object (all keys[0]) as well as the temperatureMax field OUTSIDE of the daypart object
        # will appear as null in the API after 3:00pm Local Apparent Time.
        # this addon only uses iconCode and temperature in the forecast at this time
        # so after 3pm set iconCode to the night item and temperature to 'na'
        if data['daypart'][0]['iconCode'][2 * count] == 'na':
            weathercode = WEATHER_CODES[data['daypart'][0]['iconCode'][2 * count + 1]]
        else:
            weathercode = WEATHER_CODES[data['daypart'][0]['iconCode'][2 * count]]
        set_property('Daily.%i.LongDay' % (count + 1), item)
        set_property('Daily.%i.ShortDay' % (count+1), item)
        date = datetime.datetime.fromtimestamp(data['validTimeUtc'][count])
        if DATEFORMAT[1] == 'd':
            set_property('Daily.%i.LongDate'         % (count+1), str(date.day) + ' ' + date.month)
            set_property('Daily.%i.ShortDate'        % (count+1), str(date.day) + ' ' + MONTH[date.month])
        else:
            set_property('Daily.%i.LongDate'         % (count+1), str(date.month) + ' ' + str(date.day))
            set_property('Daily.%i.ShortDate'        % (count+1), MONTH[date.month] + ' ' + str(date.day))
        set_property('Daily.%i.Outlook'              % (count+1), str(data['daypart'][0]['wxPhraseLong'][2*count]))
        set_property('Daily.%i.OutlookIcon'          % (count+1), WEATHER_ICON % weathercode)
        set_property('Daily.%i.FanartCode'           % (count+1), str(weathercode))
        if SPEEDUNIT == 'mph':
            set_property('Daily.%i.WindSpeed'        % (count+1), str(data['daypart'][0]['windSpeed'][2*count]) + ' ' + SPEEDUNIT)
        elif SPEEDUNIT == 'Beaufort':
            set_property('Daily.%i.WindSpeed'        % (count+1), KPHTOBFT(data['daypart'][0]['windSpeed'][2*count]))
        else:
            set_property('Daily.%i.WindSpeed'        % (count+1), str(data['daypart'][0]['windSpeed'][2*count]) + ' ' + SPEEDUNIT)
            set_property('Daily.%i.WindDirection'        % (count+1), str(data['daypart'][0]['windDirection'][2*count] or 0))
            set_property('Daily.%i.ShortWindDirection'   % (count+1), str(data['daypart'][0]['windDirectionCardinal'][2*count] or 0))
            set_property('Daily.%i.WindDegree'           % (count+1), str(data['daypart'][0]['windDirection'][2*count] or 0))
            set_property('Daily.%i.Humidity'             % (count+1), str(data['daypart'][0]['relativeHumidity'][2*count] or 0) + '%')
        if 'F' in TEMPUNIT:
            # null after 3 pm
            if data['daypart'][0]['temperature'][2*count] == 'na':
                set_property('Daily.%i.HighTemperature' % (count + 1), str('na'))
                set_property('Daily.%i.LowTemperature' % (count + 1), str(CELSIUStoFAHR(data['daypart'][0]['temperature'][2 * count + 1])) + TEMPUNIT)
                set_property('Daily.%i.Precipitation' % (count + 1), str(data['daypart'][0]['qpf'][2 * count]) + ' in')
            else:
                set_property('Daily.%i.HighTemperature'  % (count+1), str(CELSIUStoFAHR(data['daypart'][0]['temperature'][2*count])) + TEMPUNIT)
                set_property('Daily.%i.LowTemperature'   % (count+1), str(CELSIUStoFAHR(data['daypart'][0]['temperature'][2*count + 1])) + TEMPUNIT)
                set_property('Daily.%i.Precipitation'    % (count+1), str(data['daypart'][0]['qpf'][2*count]) + ' in')
        else:
            # null after 3 pm
            if data['daypart'][0]['temperature'][2 * count] == 'na':
                set_property('Daily.%i.HighTemperature' % (count + 1), str('na'))
                set_property('Daily.%i.LowTemperature' % (count + 1), str(data['daypart'][0]['temperature'][2 * count + 1]) + TEMPUNIT)
                set_property('Daily.%i.Precipitation' % (count + 1), str(data['daypart'][0]['qpf'][2 * count]) + ' in')
            else:
                set_property('Daily.%i.HighTemperature'  % (count+1), str(data['daypart'][0]['temperature'][2*count]) + TEMPUNIT)
                set_property('Daily.%i.LowTemperature'   % (count+1), str(data['daypart'][0]['temperature'][2*count + 1]) + TEMPUNIT)
                set_property('Daily.%i.Precipitation'    % (count+1), str(data['daypart'][0]['qpf'][2*count]) + ' in')

log('version %s started: %s' % (VERSION, sys.argv))
log('lang: %s'    % LOCALIZE)
log('speed: %s'   % SPEEDUNIT)
log('temp: %s'    % TEMPUNIT[1])
log('time: %s'    % TIMEFORMAT)
log('date: %s'    % DATEFORMAT)

set_property('WeatherProvider', ADDONNAME)
set_property('WeatherProviderLogo', xbmc.translatePath(os.path.join(CWD, 'resources', 'banner.png')))

if not API:
    log('no api key provided')
else:
    location = ADDON.getSetting('Location%s' % sys.argv[1])
    locationid = ADDON.getSetting('Location%sid' % sys.argv[1])
    locationlat = ADDON.getSetting('Location%slat' % sys.argv[1])
    locationlon = ADDON.getSetting('Location%slon' % sys.argv[1])
    if (locationid == '') and (sys.argv[1] != '1'):
        location = ADDON.getSetting('Location1')
        locationid = ADDON.getSetting('Location1id')
        locationlat = ADDON.getSetting('Location1lat')
        locationlon = ADDON.getSetting('Location1lon')
        log('trying location 1 instead')
    if not locationid == '':
        forecast(location, locationid, locationlat, locationlon)
    else:
        log('no location found')
        clear()
    refresh_locations()

log('finished')
